site/en/docs/extensions/mv3/geolocation/index.md

```diff
@@ -0,0 +1,204 @@
+});
+```
+
+Note that when you access an offscreen document, you need to include a `reason`. The `geolocation` reason was not originally available, so we specify a fallback of `DOM\_SCRAPING`, and explain in our `justification` section what the code is actually doing. This information is used by the Chrome Web Store's review process to ensure offscreen documents are being used for a valid purpose.
+
+Once you have a reference to the Offscreen Document, you can send it a message to ask for it to give you updated geolocation information.
+
+{% Label %}service\_worker.js:{% endLabel %}
+
+```js
+const OFFSCREEN_DOCUMENT_PATH = '/offscreen.html';
+let creating; // A global promise to avoid concurrency issues
+
+chrome.runtime.onMessage.addListener(handleMessages);
+
+async function getGeolocation() {
+  await setupOffscreenDocument(OFFSCREEN_DOCUMENT_PATH);
+  const geolocation = await chrome.runtime.sendMessage({
+    type: 'get-geolocation',
+    target: 'offscreen'
+  });
+  await closeOffscreenDocument();
+  return geolocation;
+}
+
+async function hasDocument() {
+  // Check all windows controlled by the service worker to see if one
+  // of them is the offscreen document with the given path
+  const offscreenUrl = chrome.runtime.getURL(OFFSCREEN_DOCUMENT_PATH);
+  const matchedClients = await clients.matchAll();
+
+  return matchedClients.some(c => c.url === offscreenUrl)
+}
+
+async function setupOffscreenDocument(path) {
+  //if we do not have a document, we are already setup and can skip
+  if (!(await hasDocument())) {
+    // create offscreen document
+    if (creating) {
+      await creating;
+    } else {
+      creating = chrome.offscreen.createDocument({
+        url: path,
+        reasons: [chrome.offscreen.Reason.GEOLOCATION || chrome.offscreen.Reason.DOM_SCRAPING],
+        justification: 'add justification for geolocation use here',
+      });
+
+      await creating;
+      creating = null;
+    }
+  }
+}
+
+async function closeOffscreenDocument() {
+  if (!(await hasDocument())) {
+    return;
+  }
+  await chrome.offscreen.closeDocument();
+}
+```
+
+So now whenever you want to get the geolocation from your service worker, you just need to call
+
+```js
+const location = await getGeolocation()
+```
+
+### Using geolocation in popup or sidepanel
+
+If you want to use geolocation within a [popup](popup) or [side panel](sidepanel), it is very straightforward. Popups and sidepanels are just web documents, and therefore have access to the normal DOM APIs. You can access `navigator.geolocation` directly. The only difference from standard web sites would be that you need to use the manifest.json `permission` field to request the `geolocation` permission. If you do not include the permission, you _will_ still have access to `navigator.geolocation`. However, any attempt to use it will cause an immediate error, the same as if the user rejected the request. You can see a sample of this [here](popup-sample).
```

Comment:
```suggestion
If you want to use geolocation within a [popup](popup) or [side panel](sidepanel), it is very straightforward. Popups and side panels are just web documents, and therefore have access to the normal DOM APIs. You can access `navigator.geolocation` directly. The only difference from standard web sites would be that you need to use the `manifest.json` `"permission"` field to request the `geolocation` permission. If you do not include the permission, you _will_ still have access to `navigator.geolocation`. However, any attempt to use it will cause an immediate error, the same as if the user rejected the request. You can see this in the [popup sample](popup-sample).
```